### PR TITLE
init mvvm structure + test authorize screen

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 75863: Content](https://edu-redmine.sun-asterisk.vn/issues/75863)
## WHAT
- init mvvm structure + test authorize screen
## Evidence (Screenshot or Video)
[Screen_recording_20240514_150928.webm](https://github.com/awesome-academy/HN_Android_2024_Image_Editor_2/assets/163242235/31e929d5-cdc7-4e64-82f4-d78d8be8a564)

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/blob/master/eng/android/coding_convention_android_kotlin.md | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(Other notes)*

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -2,7 +2,9 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
 
+    <uses-permission android:name="android.permission.INTERNET" />
     <application
+        android:name=".App"
```
Comment (daolq-2712) [inline app/src/main/AndroidManifest.xml:7]:
```suggestion
        android:name=".App"
```
set thêm chỉ support màn hình portrait e nhé!

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -13,7 +15,7 @@
         android:theme="@style/Theme.ImageEditor2"
         tools:targetApi="31">
```
Comment (daolq-2712) [inline app/src/main/AndroidManifest.xml]:
```suggestion
        tools:targetApi="33">
```

File: app/src/main/java/com/example/imageEditor2/base/BaseActivity.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import androidx.appcompat.app.AppCompatActivity
+import androidx.viewbinding.ViewBinding
+import com.example.imageEditor2.R
+import com.example.imageEditor2.base.BaseViewModel
+
+abstract class BaseActivity<VB : ViewBinding> : AppCompatActivity() {
+    private lateinit var _binding: VB
+    private lateinit var viewModel: BaseViewModel
+    val binding get() = _binding
+    private val loadingDialog by lazy { setupProgressDialog() }
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/base/BaseActivity.kt]:
```suggestion
    private lateinit var _binding: VB
     val binding get() = _binding
    private lateinit var viewModel: BaseViewModel
    
    private val loadingDialog by lazy { setupProgressDialog() }
```

File: app/src/main/java/com/example/imageEditor2/module/NetworkModule.kt [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        val httpClient = OkHttpClient.Builder()
+        httpClient.connectTimeout(TIME_OUT, TimeUnit.SECONDS)
+        httpClient.readTimeout(TIME_OUT, TimeUnit.SECONDS)
+
+        httpClient.addInterceptor { chain ->
+            val request =
+                chain.request().newBuilder()
+                    .addHeader("Content-Type", "application/json")
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/module/NetworkModule.kt]:
```suggestion
                    .addHeader(CONTENT_TYPE, "application/json")
```
Hard code & đang duplicate vs bên dưới ref: [link](https://github.com/awesome-academy/HN_Android_2024_Image_Editor_2/pull/2/files#diff-76e58641e721fa3662a0825135ab04ca38878c9cee95687629405be7b933df3dR69)

sửa tương tự vs các giá trị khác, hạn chế hard code

**Review by daolq-2712 — CHANGES_REQUESTED** [review]
Pls check my comments!

File: app/src/main/java/com/example/imageEditor2/module/NetworkModule.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    var mToken = ""
+    private const val TIME_OUT: Long = 10
+    private var mLanguage = "vi"
+
+    internal fun providePostApi(retrofit: Retrofit): Api {
+        return retrofit.create(Api::class.java)
+    }
+
+    internal fun provideRetrofitInterface(context: Context): Retrofit {
+        val mPrefs = PreferenceManager.getDefaultSharedPreferences(context)
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/module/NetworkModule.kt]:
```suggestion
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
```

Comment (daolq-2712) [reply app/src/main/java/com/example/imageEditor2/module/NetworkModule.kt]: Naming -
biến bên trong hàm thì như này e nhé!

File: app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeActivity.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                    if (request?.url.toString().contains(SIGN_OF_AUTHORIZE)) {
+                        viewModel.authorize(request?.url.toString().toAuthorizationCode())
+                    }
+                    return false
+                }
+            }
+    }
+
+    override fun initListener() {
+    }
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeActivity.kt]:
```suggestion
    override fun initListener() {
       /*No-op*/
    }
```

File: app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                AuthorizeRequest(
+                    clientId = ACCESS_KEY,
+                    clientSecret = SECRET_KEY,
+                    redirectUri = REDIRECT_URI,
+                    code = authorizationCode,
+                )
+            authorizeRepository.authorize(authorizeRequest).onStart { showLoading() }
+                .catch { handleApiError(it) }
+                .onCompletion { hideLoading() }
+                .collect {
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeViewModel.kt:34]: Chuyển về main thread trc khi call collect, như này đang ở IO thread

**Review by daolq-2712 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeViewModel.kt (hunk 7 above)
Comment (ghost) [reply app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeViewModel.kt:34]: theo em tìm hiểu thì collect đang được chạy trên coroutineContext triển khai nó(viewModelScope.launch: MainThread) ạ,

Comment (daolq-2712) [reply app/src/main/java/com/example/imageEditor2/ui/authorize/AuthorizeViewModel.kt:34]: @thold-3445 
OK, khi em run lên sẽ gặp lỗi, sửa sau cũng dc :))

**Review by daolq-2712 — APPROVED** [review]
OK!

**Pull request merged**